docs/available-filters/envelope-encryption.adoc

```diff
@@ -0,0 +1,296 @@
+5. The filter forwards the modified fetch response onto the Kafka Client. The client has no knowledge that the record was encrypted.
+
+The entire process is transparent from the point of view of both Kafka Client and Kafka Broker.  Neither are
+aware that the records are being encrypted.  Neither client nor broker have any access to the encryption keys.
+
+The filter encrypts the records using symmetric encryption keys.  The encryption technique employed is
+known as *Envelope Encryption*, which is technique suited for encrypting large volumes of data in a efficient manner.
+Envelope Encryption is described in the next section.
+
+The filter integrates with a Key Management Service (KMS).  It is the Key Management Service (KMS) that has
+ultimate responsibility for the safe storage of key material.  The role of the KMS is discussed later.
+
+=== Envelope Encryption
+
+Envelope Encryption is a technique described by https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-57pt1r5.pdf[NIST
+Special Publication 800-57 Part 1 Revision 5]. It is the practice of encrypting the data with a data encryption key (DEK),
+then wrapping (encrypting) the DEK with a key encryption key (KEK).  In this section we discuss how Envelope Encryption is
+applied by the filter to encrypt Kafka records.  For more detail, refer to the {design-doc}[design].
+
+On the produce path, the filter is encrypting records.  The filter uses a selector to determine KEK is use. It then
+requests that the KMS generates a DEK for the KEK.  It is the DEK that is used to encrypt the record.  The original record
+in the produce request is replaced by a *cipher record* which comprises the encrypted record, the encrypted DEK and some other
+data.
+
+On the fetch path, the filter is decrypting records.  The filter receives the *cipher record* from the Kafka Broker. The
+filter reverses process used to  construct the *cipher record* and uses the KMS to decrypt DEK.  The decrypted DEK is
+used to decrypt the encrypted record.  The cipher record in the fetch response is then replaced with the decrypted record.
+
+On the produce path, the filter employs a DEK reuse strategy. This means that records sent to same topic will be encrypted
+using the same DEK (until a time-out or number of encryption limit is reached, whichever occurs first).  This prevents
+key exhaustion and prevents excessive interactions with the KMS.
+
+On the consume path, the filter employs an LRU strategy to keep recently encountered DEKs decrypted in memory. This
+prevents excessive interactions with the KMS.
+
+=== Role of the KMS
+
+The Key Management Service (KMS) is the *secure repository* for Key Management Keys (KEKs). The key material of the KEK
+*never* leaves the confines of the KMS.  The KMS exposes cryptographic functions that operate against the KEKs stored
+within it.
+
+The KMS is also the *source* of Data Encryption Keys (DEKs).  When the KMS generates a DEK for a given KEK, it returns
```

Comment:
Should we explain that DEKs are not stored in the KMS? We use the KMS to generate the DEK but from then on it lives with the data.

Reply:
In the section before I already have:

"The original record in the produce request is replaced by a *cipher record* which comprises the encrypted record, the encrypted DEK and some other data."

so we were already explaining it, but I've added an extra sentence to this section too.